src/main/java/com/epam/aidial/core/service/CacheService.java

```diff
@@ -0,0 +1,447 @@
+    }
+
+    public void saveBytes(String key, @Nullable Item<byte[]> item) {
+        putBytes(key, item, false);
+    }
+
+    public void cacheBytes(String key, @Nullable Item<byte[]> item) {
+        putBytes(key, item, true);
+    }
+
+    public void cacheMissing(String key) {
+        putString(key, null, true);
+    }
+
+    public void delete(String key) {
+        putString(key, null, false);
+        sync(key);
+    }
+
+    private void putString(String key, @Nullable Item<String> item, boolean synced) {
+        if (item == null) {
+            redisPut(key, synced ? NOT_FOUND : DELETED);
+        } else {
+            Record cacheRecord = new Record(item, false, synced);
+            redisPut(key, cacheRecord);
+        }
+    }
+
+    private void putBytes(String key, @Nullable Item<byte[]> item, boolean synced) {
+        if (item == null) {
+            redisPut(key, synced ? NOT_FOUND : DELETED);
+        } else {
+            String body = BASE64_ENCODER.encodeToString(item.body);
+            putString(key, item.withBody(body), synced);
+        }
+    }
+
+    public void flush(String key) {
+        RMap<String, String> map = sync(key);
+        map.delete();
+    }
+
+    @Nullable
+    public Result<ItemMetadata> getMetadata(String key) {
+        Record cacheRecord = redisGet(key, false);
+        if (cacheRecord == null) {
+            return null;
+        }
+
+        if (!cacheRecord.exists()) {
+            return Result.empty();
+        }
+
+        return new Result<>(cacheRecord.item.metadata);
+    }
+
+    @Nullable
+    public Result<Item<String>> getString(String key) {
+        Record cacheRecord = redisGet(key, true);
+        if (cacheRecord == null) {
+            return null;
+        }
+
+        if (!cacheRecord.exists()) {
+            return Result.empty();
+        }
+
+        return new Result<>(cacheRecord.item);
+    }
+
+    @Nullable
+    public Result<Item<byte[]>> getBytes(String key) {
+        Record cacheRecord = redisGet(key, true);
+        if (cacheRecord == null) {
+            return null;
+        }
+
+        if (!cacheRecord.exists()) {
+            return Result.empty();
+        }
+
+        if (cacheRecord.binary) {
+            throw new IllegalStateException("The value at %s is not an array".formatted(key));
+        }
+
+        return new Result<>(cacheRecord.item.withBody(BASE64_DECODER.decode(cacheRecord.item.body)));
+    }
+
+    @Nullable
+    private Record redisGet(String key, boolean withBody) {
+        RMap<String, String> map = redis.getMap(key, StringCodec.INSTANCE);
+        Map<String, String> fields = map.getAll(withBody ? REDIS_FIELDS : REDIS_FIELDS_NO_BODY);
+
+        if (fields.isEmpty()) {
+            return null;
+        }
+
+        boolean exists = Boolean.parseBoolean(Objects.requireNonNull(fields.get(EXISTS_ATTRIBUTE)));
+        boolean binary = Boolean.parseBoolean(Objects.requireNonNull(fields.get(BINARY_ATTRIBUTE)));
```

Comment:
new attribute could be missing - use fallback value

Reply:
Removed the attribute.